File: scripts/nm_health.py
#!/usr/bin/env python3
# ─────────────────────────────────────────────────────────────────────────────
# NEURU — Predictive Link & Hardware Health COLLECTOR.
#
# Standalone, isolated from nm_poller.py on purpose (a bug here must NOT break
# core polling). One-shot by default (cron-friendly); pass --loop to daemonize.
#
# Collects per node + interface:
#   ETHERNET (universal, standard OIDs):
#     in_err_rate   ifInErrors    .1.3.6.1.2.1.2.2.1.14   (Counter → /sec rate)
#     out_err_rate  ifOutErrors   .1.3.6.1.2.1.2.2.1.20
#     discard_rate  ifInDiscards+ifOutDiscards (.13 + .19)
#     fcs_rate      dot3StatsFCSErrors  .1.3.6.1.2.1.10.7.2.1.3
#   OPTICAL (SFP DOM, vendor-specific — driven by nm_health_optical_oids, the
#     rows the user ENABLED for their gear): rx_dbm / tx_dbm / temp_c / bias_ma / volt_v
#
# Writes gauges/rates to nm_health_samples; nm_health.php (PHP engine) then trends
# them and predicts days-to-failure. Counters' previous values live in
# nm_health_counters for delta computation.
# ─────────────────────────────────────────────────────────────────────────────
import sys, re, time, subprocess
from datetime import datetime
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from nm_db_config import DB
import mysql.connector

INTERVAL = 300  # seconds between passes in --loop mode
MAX32 = 0xFFFFFFFF

# Realistic per-second ceilings for ethernet error/discard rates. Anything above
# these is not a real degradation signal: some devices (e.g. certain Cisco APs)
# report a packet/octet counter in the ifInErrors OID, which yields hundreds of
# thousands of "errors/s" and triggers false "Degrading" alarms. Drop those — a real
# failing link sits in the 0.001–100/s range, so these limits never hide a true fault.
PLAUSIBLE_MAX = {
    'in_err_rate':  5000.0,
    'out_err_rate': 5000.0,
    'fcs_rate':     5000.0,
    'discard_rate': 5000.0,
}

ETH_OIDS = {
    'in_err':      '.1.3.6.1.2.1.2.2.1.14',
    'out_err':     '.1.3.6.1.2.1.2.2.1.20',
    'in_disc':     '.1.3.6.1.2.1.2.2.1.13',
    'out_disc':    '.1.3.6.1.2.1.2.2.1.19',
    'fcs':         '.1.3.6.1.2.1.10.7.2.1.3',
}

def log(m): print(f"[{datetime.now():%H:%M:%S}] {m}", flush=True)
def get_db(): return mysql.connector.connect(**DB, autocommit=False)

def parse_num(s):
    if s is None: return None
    m = re.search(r'-?[\d.]+', str(s))
    return float(m.group()) if m else None

def snmp_walk(ip, community, ver, oid, timeout=3):
    v = '1' if ver == 'v1' else '2c'
    base = oid.lstrip('.')
    try:
        r = subprocess.run(['/usr/bin/snmpwalk', f'-v{v}', '-c', community, '-Oqn',
                            '-t', str(timeout), '-r', '1', ip, oid],
                           capture_output=True, text=True, timeout=timeout * 3)
        out = []
        for line in r.stdout.strip().split('\n'):
            if not line.strip(): continue
            parts = line.split(None, 1)
            if len(parts) != 2: continue
            full = parts[0].strip(); val = parts[1].strip().strip('"')
            stripped = full.lstrip('.')
            suffix = stripped[len(base):].lstrip('.') if stripped.startswith(base) else full.split('.')[-1]
            out.append((suffix, val))
        return out
    except Exception as e:
        log(f"    snmpwalk {oid} on {ip}: {e}")
        return []

# os_icon → optical vendor_key (mirrors nm_cm_guess_vendor)
def guess_vendor(os_icon):
    s = (os_icon or '').lower()
    if s in ('routeros', 'mikrotik'): return 'mikrotik'
    if s == 'cisco': return 'cisco'
    return 'generic'

def ensure(cur):
    cur.execute("""CREATE TABLE IF NOT EXISTS nm_health_samples (
        id BIGINT AUTO_INCREMENT PRIMARY KEY, node_id INT NOT NULL,
        kind VARCHAR(12) NOT NULL DEFAULT 'eth', entity VARCHAR(120) NOT NULL,
        metric VARCHAR(24) NOT NULL, value DOUBLE DEFAULT NULL, recorded_at DATETIME NOT NULL,
        KEY idx_series (node_id, entity, metric, recorded_at), KEY idx_time (recorded_at)
    ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4""")
    cur.execute("""CREATE TABLE IF NOT EXISTS nm_health_counters (
        node_id INT NOT NULL, if_index INT NOT NULL, metric VARCHAR(24) NOT NULL,
        counter BIGINT NOT NULL, polled_at DATETIME NOT NULL,
        PRIMARY KEY (node_id, if_index, metric)
    ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4""")
    cur.execute("""CREATE TABLE IF NOT EXISTS nm_health_optical_oids (
        id INT AUTO_INCREMENT PRIMARY KEY, vendor_key VARCHAR(40) NOT NULL DEFAULT 'generic',
        node_id INT DEFAULT NULL, label VARCHAR(80) NOT NULL DEFAULT '', metric VARCHAR(24) NOT NULL,
        oid VARCHAR(160) NOT NULL, scale DOUBLE NOT NULL DEFAULT 1, enabled TINYINT(1) NOT NULL DEFAULT 0,
        created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP, KEY idx_vendor (vendor_key), KEY idx_node (node_id)
    ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4""")

def rate(cur, node_id, if_index, metric, counter, now_dt, now_s):
    """Compute per-second rate vs the stored previous counter; upsert the counter."""
    cur.execute("SELECT counter, polled_at FROM nm_health_counters WHERE node_id=%s AND if_index=%s AND metric=%s",
                (node_id, if_index, metric))
    row = cur.fetchone()
    out = None
    if row is not None:
        prev_c, prev_t = int(row[0]), row[1]
        elapsed = (now_dt - prev_t).total_seconds()
        if elapsed > 0:
            d = counter - prev_c
            if d < 0: d = (counter - prev_c) % (MAX32 + 1)   # counter wrap
            r = d / elapsed
            # Drop physically-impossible rates (broken/non-standard counters, e.g. an AP
            # reporting a packet counter in the error OID) so they never alarm.
            ceil = PLAUSIBLE_MAX.get(metric, 1e6)
            if r <= ceil:
                out = round(r, 4)
    cur.execute("""INSERT INTO nm_health_counters (node_id, if_index, metric, counter, polled_at)
                   VALUES (%s,%s,%s,%s,%s) ON DUPLICATE KEY UPDATE counter=%s, polled_at=%s""",
                (node_id, if_index, metric, counter, now_s, counter, now_s))
    return out

def collect_node(cur, node, now_dt, now_s):
    node_id, ip, community, ver, os_icon = node
    if not ip or not community: return 0
    # if_index → interface name
    cur.execute("SELECT if_index, if_name FROM nm_interfaces WHERE node_id=%s AND if_index IS NOT NULL AND show_graph=1", (node_id,))
    names = {int(r[0]): (r[1] or f"if{r[0]}") for r in cur.fetchall()}
    if not names:
        return 0
    written = 0
    # ── Ethernet error/discard/FCS counters → rates ──
    raw = {k: {} for k in ETH_OIDS}
    for k, oid in ETH_OIDS.items():
        for suffix, val in snmp_walk(ip, community, ver, oid):
            v = parse_num(val)
            idx = suffix.split('.')[-1] if suffix else None
            if v is not None and idx is not None and idx.isdigit():
                raw[k][int(idx)] = int(v)
    samples = []
    for idx, ifname in names.items():
        defs = [
            ('in_err_rate',  raw['in_err'].get(idx)),
            ('out_err_rate', raw['out_err'].get(idx)),
            ('fcs_rate',     raw['fcs'].get(idx)),
        ]
        # discards = in + out
        di, do = raw['in_disc'].get(idx), raw['out_disc'].get(idx)
        if di is not None or do is not None:
            defs.append(('discard_rate', (di or 0) + (do or 0)))
        for metric, counter in defs:
            if counter is None: continue
            r = rate(cur, node_id, idx, metric, counter, now_dt, now_s)
            if r is not None:
                samples.append((node_id, 'eth', ifname, metric, r, now_s))
    # ── Optical DOM (only enabled OID rows for this node / its vendor) ──
    vendor = guess_vendor(os_icon)
    cur.execute("""SELECT label, metric, oid, scale FROM nm_health_optical_oids
                   WHERE enabled=1 AND (node_id=%s OR (node_id IS NULL AND vendor_key IN (%s,'generic','entity_sensor')))""",
                (node_id, vendor))
    for label, metric, oid, scale in cur.fetchall():
        for suffix, val in snmp_walk(ip, community, ver, oid):
            v = parse_num(val)
            if v is None: continue
            ent = f"{label} #{suffix}" if suffix else label
            samples.append((node_id, 'optical', ent[:120], metric, v * float(scale), now_s))
    if samples:
        cur.executemany("INSERT INTO nm_health_samples (node_id,kind,entity,metric,value,recorded_at) VALUES (%s,%s,%s,%s,%s,%s)", samples)
        written = len(samples)
    return written

def run_once():
    db = get_db(); cur = db.cursor()
    ensure(cur); db.commit()
    cur.execute("""SELECT id, COALESCE(ip_address,''), COALESCE(snmp_community,''),
                          COALESCE(snmp_version,'v2c'), COALESCE(os_icon,'')
                   FROM nm_nodes
                   WHERE COALESCE(monitor_type,'snmp')='snmp' AND snmp_community IS NOT NULL AND snmp_community<>''""")
    nodes = cur.fetchall()
    now_dt = datetime.now(); now_s = now_dt.strftime('%Y-%m-%d %H:%M:%S')
    total = 0
    for node in nodes:
        try:
            n = collect_node(cur, node, now_dt, now_s)
            db.commit(); total += n
        except Exception as e:
            db.rollback(); log(f"  node {node[0]} error: {e}")
    # prune stale counters (no longer polled)
    try:
        cur.execute("DELETE FROM nm_health_counters WHERE polled_at < DATE_SUB(NOW(), INTERVAL 2 DAY)")
        db.commit()
    except Exception:
        db.rollback()
    log(f"health collect: {total} sample(s) across {len(nodes)} node(s)")
    cur.close(); db.close()

def main():
    loop = '--loop' in sys.argv
    while True:
        try:
            run_once()
        except Exception as e:
            log(f"run error: {e}")
        if not loop:
            break
        time.sleep(INTERVAL)

if __name__ == '__main__':
    main()

File: nm_health.php
<?php
// ─────────────────────────────────────────────────────────────────────────────
// NEURU — Predictive Link & Hardware Health engine.
//
//   Proactive (not reactive) degradation forecasting for BOTH:
//     · OPTICAL fiber (SFP DOM/DDM): Rx/Tx power dBm, laser bias mA, temp °C, Vcc
//     · ETHERNET / copper: error-counter RATES (Rx/Tx errors, discards, CRC/FCS)
//
//   The Python collector (scripts/nm_health.py) samples these per node+interface
//   into nm_health_samples. This engine fits a least-squares trend over a window
//   and predicts WHEN a metric will cross its critical threshold — e.g. "the SFP
//   Rx power on Node X is dropping 0.4 dBm/day and will hit −18 dBm in ~11 days —
//   schedule field maintenance". Same linear-regression idea as nm_sla.php's
//   capacity forecast, generalized to any health metric + direction.
//
//   Page gate = permission key 'health'.  See also [[netmon-sla]] nm_cap_forecast.
// ─────────────────────────────────────────────────────────────────────────────
require_once __DIR__ . '/nm_audit.php';

if (!function_exists('nm_health_ensure')) {

    // ── Metric catalog: label, unit, kind, trend direction, thresholds ───────
    // dir 'fall' = degrades as it drops (optical power); 'rise' = degrades as it
    // climbs (errors, temperature). 'crit'/'warn' are the thresholds we forecast to.
    function nm_health_metrics(): array {
        return [
            'rx_dbm'       => ['label'=>'Optical Rx power','unit'=>'dBm','kind'=>'optical','dir'=>'fall','warn'=>-14,'crit'=>-18],
            'tx_dbm'       => ['label'=>'Optical Tx power','unit'=>'dBm','kind'=>'optical','dir'=>'fall','warn'=>-8,'crit'=>-11],
            'temp_c'       => ['label'=>'SFP temperature','unit'=>'°C','kind'=>'optical','dir'=>'rise','warn'=>70,'crit'=>80],
            'bias_ma'      => ['label'=>'Laser bias current','unit'=>'mA','kind'=>'optical','dir'=>'rise','warn'=>11,'crit'=>13],
            'volt_v'       => ['label'=>'SFP supply voltage','unit'=>'V','kind'=>'optical','dir'=>'rise','warn'=>3.6,'crit'=>3.8],
            'in_err_rate'  => ['label'=>'Rx error rate','unit'=>'err/s','kind'=>'eth','dir'=>'rise','warn'=>0.5,'crit'=>5],
            'out_err_rate' => ['label'=>'Tx error rate','unit'=>'err/s','kind'=>'eth','dir'=>'rise','warn'=>0.5,'crit'=>5],
            'fcs_rate'     => ['label'=>'CRC/FCS error rate','unit'=>'err/s','kind'=>'eth','dir'=>'rise','warn'=>0.5,'crit'=>5],
            'discard_rate' => ['label'=>'Discard rate','unit'=>'pkt/s','kind'=>'eth','dir'=>'rise','warn'=>1,'crit'=>20],
        ];
    }
    function nm_health_metric($m): ?array { $a=nm_health_metrics(); return $a[$m]??null; }

    function nm_health_ensure($conn): void {
        if (!($conn instanceof mysqli)) return;
        static $done=false; if($done) return; $done=true;
        $conn->query("CREATE TABLE IF NOT EXISTS nm_health_samples (
            id BIGINT AUTO_INCREMENT PRIMARY KEY,
            node_id INT NOT NULL,
            kind VARCHAR(12) NOT NULL DEFAULT 'eth',
            entity VARCHAR(120) NOT NULL,
            metric VARCHAR(24) NOT NULL,
            value DOUBLE DEFAULT NULL,
            recorded_at DATETIME NOT NULL,
            KEY idx_series (node_id, entity, metric, recorded_at),
            KEY idx_time (recorded_at)
        ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4");
        $conn->query("CREATE TABLE IF NOT EXISTS nm_health_predictions (
            id BIGINT AUTO_INCREMENT PRIMARY KEY,
            node_id INT NOT NULL,
            entity VARCHAR(120) NOT NULL,
            metric VARCHAR(24) NOT NULL,
            kind VARCHAR(12) NOT NULL DEFAULT 'eth',
            severity VARCHAR(8) NOT NULL DEFAULT 'ok',
            direction VARCHAR(4) NOT NULL DEFAULT 'rise',
            current_val DOUBLE DEFAULT NULL,
            threshold DOUBLE DEFAULT NULL,
            slope_per_day DOUBLE DEFAULT NULL,
            eta_days DOUBLE DEFAULT NULL,
            samples INT DEFAULT 0,
            detail VARCHAR(400) DEFAULT NULL,
            status VARCHAR(10) NOT NULL DEFAULT 'open',
            first_seen DATETIME DEFAULT NULL,
            updated_at DATETIME DEFAULT NULL,
            UNIQUE KEY uniq_series (node_id, entity, metric),
            KEY idx_status (status, severity)
        ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4");
        // Make sure the admin role can reach the page (idempotent).
        @$conn->query("INSERT INTO role_profiles (role_name,button_key,enabled)
            SELECT 'admin','health',1 FROM DUAL
            WHERE NOT EXISTS (SELECT 1 FROM role_profiles WHERE role_name='admin' AND button_key='health')");
        // Per-node / per-vendor optical OID config for the collector.
        $conn->query("CREATE TABLE IF NOT EXISTS nm_health_optical_oids (
            id INT AUTO_INCREMENT PRIMARY KEY,
            vendor_key VARCHAR(40) NOT NULL DEFAULT 'generic',
            node_id INT DEFAULT NULL,
            label VARCHAR(80) NOT NULL DEFAULT '',
            metric VARCHAR(24) NOT NULL,
            oid VARCHAR(160) NOT NULL,
            scale DOUBLE NOT NULL DEFAULT 1,
            enabled TINYINT(1) NOT NULL DEFAULT 0,
            created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
            KEY idx_vendor (vendor_key), KEY idx_node (node_id)
        ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4");
        // Seed common SFP DOM OID templates (disabled by default — the user enables
        // the ones that match their gear). ENTITY-SENSOR is generic; vendors vary.
        $r=$conn->query("SELECT COUNT(*) c FROM nm_health_optical_oids");
        if ($r && (int)$r->fetch_assoc()['c']===0) {
            $seed = [
                // vendor, label, metric, oid (base — collector walks the subtree), scale
                ['entity_sensor','ENTITY-SENSOR entPhySensorValue','rx_dbm','.1.3.6.1.2.1.99.1.1.1.4',1],
                ['mikrotik','MikroTik SFP Rx power (×0.01 dBm? verify)','rx_dbm','.1.3.6.1.4.1.14988.1.1.19.1.1.7',0.01],
                ['mikrotik','MikroTik SFP Tx power','tx_dbm','.1.3.6.1.4.1.14988.1.1.19.1.1.8',0.01],
                ['mikrotik','MikroTik SFP temperature','temp_c','.1.3.6.1.4.1.14988.1.1.19.1.1.3',0.001],
                ['mikrotik','MikroTik SFP Tx bias','bias_ma','.1.3.6.1.4.1.14988.1.1.19.1.1.6',0.001],
                ['mikrotik','MikroTik SFP supply voltage','volt_v','.1.3.6.1.4.1.14988.1.1.19.1.1.5',0.001],
                ['cisco','Cisco entSensorValue (DOM)','rx_dbm','.1.3.6.1.4.1.9.9.91.1.1.1.1.4',1],
            ];
            $st=$conn->prepare("INSERT INTO nm_health_optical_oids (vendor_key,label,metric,oid,scale,enabled) VALUES (?,?,?,?,?,0)");
            foreach($seed as $s){ $st->bind_param('sssds',$s[0],$s[1],$s[2],$s[3],$s[4]); $st->execute(); }
            $st->close();
        }
    }

    function nm_health_setting($conn,$k,$d){ $r=$conn->query("SELECT setting_val FROM nm_settings WHERE setting_key='".$conn->real_escape_string($k)."' LIMIT 1"); return $r&&($x=$r->fetch_row())?$x[0]:$d; }

    // ── Excluded links — operator-muted interfaces that report garbage counters ──
    // Some devices (e.g. certain Cisco APs) put a packet/octet counter in the error
    // OID, so a link "degrades" forever. The plausibility ceiling in nm_health.py drops
    // the worst spikes; this lets an operator permanently mute a specific node+interface.
    // Stored in nm_settings.health_exclude as a JSON array of "node_id|entity" keys.
    function nm_health_excluded($conn): array {
        $j=nm_health_setting($conn,'health_exclude','[]'); $a=json_decode((string)$j,true);
        if(!is_array($a)) return [];
        $set=[]; foreach($a as $k){ if(is_string($k)) $set[$k]=1; } return $set;
    }
    function nm_health_exclude_save($conn,array $set): void {
        $list=array_values(array_keys($set));
        $val=$conn->real_escape_string(json_encode($list));
        $conn->query("INSERT INTO nm_settings (setting_key,setting_val) VALUES ('health_exclude','$val')
            ON DUPLICATE KEY UPDATE setting_val=VALUES(setting_val)");
    }
    function nm_health_exclude_add($conn,int $node,string $entity): void {
        $set=nm_health_excluded($conn); $set[$node.'|'.$entity]=1; nm_health_exclude_save($conn,$set);
        // drop any open prediction + samples for the muted link so it stops alarming now
        $st=$conn->prepare("UPDATE nm_health_predictions SET status='cleared',severity='ok',updated_at=NOW() WHERE node_id=? AND entity=?");
        $st->bind_param('is',$node,$entity); $st->execute(); $st->close();
    }
    function nm_health_exclude_remove($conn,int $node,string $entity): void {
        $set=nm_health_excluded($conn); unset($set[$node.'|'.$entity]); nm_health_exclude_save($conn,$set);
    }

    // ── Generic least-squares forecaster ─────────────────────────────────────
    // $series = [[hoursAgoFloat (x), value (y)], …] where x increases with time.
    // Returns slope/day, recent current value, and ETA-days to $threshold given the
    // degradation $dir ('rise'|'fall'). eta=null if not trending toward failure.
    function nm_health_forecast(array $series, float $threshold, string $dir): array {
        $n=count($series);
        if($n<3) return ['n'=>$n,'slope_day'=>0.0,'current'=>null,'eta'=>null,'span_days'=>0.0];
        // ── Winsorize y to [p05,p95] so a lone spike/dropout can't fake a trend ──
        // (a real degradation is a rising/falling BASELINE — many points move, so the
        //  percentiles move with it; only isolated outliers get clipped.)
        $ys=array_map(fn($p)=>$p[1],$series); sort($ys);
        $pct=function(float $q) use ($ys,$n){ return $ys[max(0,min($n-1,(int)round($q*($n-1))))]; };
        $lo=$pct(0.05); $hi=$pct(0.95);
        $sx=$sy=$sxx=$sxy=0.0;
        foreach($series as $p){ $x=$p[0]; $y=max($lo,min($p[1],$hi)); $sx+=$x;$sy+=$y;$sxx+=$x*$x;$sxy+=$x*$y; }
        $den=$n*$sxx-$sx*$sx;
        $slopeH = $den!=0 ? ($n*$sxy-$sx*$sy)/$den : 0.0;   // per hour (on winsorized data)
        $slopeDay=$slopeH*24;
        // current = MEDIAN of the most-recent ~20% of points (a mean is skewed by one spike)
        $tail=max(3,(int)round($n*0.2)); $recent=array_slice($series,-$tail);
        $rv=array_map(fn($p)=>$p[1],$recent); sort($rv); $m=count($rv);
        $cur = ($m%2) ? $rv[intdiv($m,2)] : ($rv[(int)($m/2)-1]+$rv[(int)($m/2)])/2.0;
        // Need a real span of history before trusting an ETA — don't forecast weeks from a
        // day or two of data. Below the floor we still report current/slope but eta=null.
        $spanDays=($series[$n-1][0]-$series[0][0])/24.0;
        $minSpan=2.0;
        $eta=null;
        if($spanDays>=$minSpan){
            if($dir==='rise' && $slopeDay>1e-6 && $cur<$threshold)  $eta=round(($threshold-$cur)/$slopeDay,1);
            if($dir==='fall' && $slopeDay<-1e-6 && $cur>$threshold) $eta=round(($cur-$threshold)/(-$slopeDay),1);
            if($eta!==null && $eta>3650) $eta=null;   // >10y out = effectively flat, not a forecast
        }
        return ['n'=>$n,'slope_day'=>round($slopeDay,4),'current'=>round($cur,3),'eta'=>$eta,'span_days'=>round($spanDays,2)];
    }

    // Pull a metric series for one (node,entity,metric) over the window (days).
    function nm_health_series($conn,int $node,string $entity,string $metric,int $windowDays): array {
        $st=$conn->prepare("SELECT UNIX_TIMESTAMP(recorded_at) t, value v FROM nm_health_samples
            WHERE node_id=? AND entity=? AND metric=? AND recorded_at>=DATE_SUB(NOW(),INTERVAL ? DAY) AND value IS NOT NULL
            ORDER BY recorded_at ASC");
        $st->bind_param('issi',$node,$entity,$metric,$windowDays); $st->execute();
        $r=$st->get_result(); $rows=[]; while($x=$r->fetch_assoc()) $rows[]=$x; $st->close();
        if(!$rows) return [];
        $t0=(float)$rows[0]['t'];
        $out=[]; foreach($rows as $x){ $out[]=[((float)$x['t']-$t0)/3600.0, (float)$x['v']]; }
        return $out;
    }

    // ── The evaluation pass: forecast every series, upsert predictions ───────
    // Returns ['scanned','warn','crit','new_crit'=>[...]]. Run from cron_health.php.
    function nm_health_evaluate($conn): array {
        nm_health_ensure($conn);
        $window = (int)nm_health_setting($conn,'health_window_days','7') ?: 7;
        $minN   = (int)nm_health_setting($conn,'health_min_samples','12') ?: 12;
        $etaWarn= (float)nm_health_setting($conn,'health_eta_warn_days','45') ?: 45;
        $etaCrit= (float)nm_health_setting($conn,'health_eta_crit_days','14') ?: 14;
        $metas=nm_health_metrics();
        $nodeNames=[]; $nr=$conn->query("SELECT id,display_name FROM nm_nodes"); while($nr&&$x=$nr->fetch_assoc()) $nodeNames[(int)$x['id']]=$x['display_name'];

        // distinct active series in the window
        $series=$conn->query("SELECT DISTINCT node_id,entity,metric FROM nm_health_samples
            WHERE recorded_at>=DATE_SUB(NOW(),INTERVAL {$window} DAY)");
        $excluded=nm_health_excluded($conn);
        $scanned=0;$warn=0;$crit=0;$newCrit=[]; $now=date('Y-m-d H:i:s');
        while($series && $s=$series->fetch_assoc()){
            $node=(int)$s['node_id']; $entity=$s['entity']; $metric=$s['metric'];
            if(isset($excluded[$node.'|'.$entity])) continue;   // operator-muted link
            $meta=$metas[$metric]??null; if(!$meta) continue;
            $data=nm_health_series($conn,$node,$entity,$metric,$window);
            if(count($data)<$minN) continue;
            $scanned++;
            $f=nm_health_forecast($data,(float)$meta['crit'],$meta['dir']);
            $cur=$f['current']; $eta=$f['eta'];
            // severity: already past crit threshold OR forecast to cross it soon.
            $pastCrit = $meta['dir']==='rise' ? ($cur>=$meta['crit']) : ($cur<=$meta['crit']);
            $pastWarn = $meta['dir']==='rise' ? ($cur>=$meta['warn']) : ($cur<=$meta['warn']);
            $sev='ok';
            if($pastCrit || ($eta!==null && $eta<=$etaCrit)) $sev='crit';
            elseif($pastWarn || ($eta!==null && $eta<=$etaWarn)) $sev='warn';

            // build a human detail line
            $nn=$nodeNames[$node]??("node ".$node);
            if($eta!==null){
                $detail=sprintf('%s on %s %s is %s %.3f→%g %s at %s%g/day — reaches critical in ~%g days.',
                    $meta['label'],$nn,$entity,($meta['dir']==='fall'?'dropping':'rising'),
                    $cur,$meta['crit'],$meta['unit'],($f['slope_day']>=0?'+':''),$f['slope_day'],$eta);
            } else {
                $detail=sprintf('%s on %s %s = %g %s (trend %s%g/day).',
                    $meta['label'],$nn,$entity,$cur,$meta['unit'],($f['slope_day']>=0?'+':''),$f['slope_day']);
            }

            // was it crit before?
            $prevSev='';
            $pg=$conn->prepare("SELECT severity FROM nm_health_predictions WHERE node_id=? AND entity=? AND metric=? LIMIT 1");
            $pg->bind_param('iss',$node,$entity,$metric); $pg->execute();
            if($pr=$pg->get_result()->fetch_assoc()) $prevSev=$pr['severity']; $pg->close();

            $etaVal = $eta===null ? null : $eta;
            $st=$conn->prepare("INSERT INTO nm_health_predictions
                (node_id,entity,metric,kind,severity,direction,current_val,threshold,slope_per_day,eta_days,samples,detail,status,first_seen,updated_at)
                VALUES (?,?,?,?,?,?,?,?,?,?,?,?, 'open', ?, ?)
                ON DUPLICATE KEY UPDATE severity=VALUES(severity),current_val=VALUES(current_val),threshold=VALUES(threshold),
                  slope_per_day=VALUES(slope_per_day),eta_days=VALUES(eta_days),samples=VALUES(samples),detail=VALUES(detail),
                  status='open',updated_at=VALUES(updated_at)");
            $kind=$meta['kind']; $thr=(float)$meta['crit']; $slope=$f['slope_day']; $npts=$f['n'];
            $st->bind_param('isssssddddisss',$node,$entity,$metric,$kind,$sev,$meta['dir'],$cur,$thr,$slope,$etaVal,$npts,$detail,$now,$now);
            $st->execute(); $st->close();

            if($sev==='crit'){ $crit++; if($prevSev!=='crit') $newCrit[]=['node'=>$nn,'entity'=>$entity,'metric'=>$metric,'detail'=>$detail]; }
            elseif($sev==='warn'){ $warn++; }
            else { // recovered → mark cleared
                $cl=$conn->prepare("UPDATE nm_health_predictions SET status='cleared',severity='ok',updated_at=? WHERE node_id=? AND entity=? AND metric=?");
                $cl->bind_param('siss',$now,$node,$entity,$metric); $cl->execute(); $cl->close();
            }
        }
        // retention prune on samples
        $ret=(int)nm_health_setting($conn,'health_retention_days','45') ?: 45;
        $conn->query("DELETE FROM nm_health_samples WHERE recorded_at < DATE_SUB(NOW(),INTERVAL {$ret} DAY)");
        // Notification Center: each newly-critical forecast is a predictive alert.
        if ($newCrit) {
            if (!function_exists('nm_notify_event')) { @include_once __DIR__.'/nm_notify.php'; }
            if (function_exists('nm_notify_event')) foreach ($newCrit as $nc) {
                @nm_notify_event($conn,'predictive','critical',
                    "Predicted failure: {$nc['metric']} on {$nc['node']} {$nc['entity']}",
                    $nc['detail'], ['entity'=>$nc['node'].' · '.$nc['entity'],'source'=>'health']);
            }
        }
        return ['scanned'=>$scanned,'warn'=>$warn,'crit'=>$crit,'new_crit'=>$newCrit];
    }

    // Open predictions for the UI / dashboard, worst first.
    function nm_health_list($conn,string $sev=''): array {
        nm_health_ensure($conn);
        $w="status='open' AND severity IN ('warn','crit')";
        if($sev==='crit') $w="status='open' AND severity='crit'";
        $rows=[]; $r=$conn->query("SELECT * FROM nm_health_predictions WHERE {$w}
            ORDER BY FIELD(severity,'crit','warn') , (eta_days IS NULL), eta_days ASC, ABS(slope_per_day) DESC LIMIT 200");
        while($r&&$x=$r->fetch_assoc()) $rows[]=$x;
        return $rows;
    }
    function nm_health_counts($conn): array {
        nm_health_ensure($conn);
        $out=['crit'=>0,'warn'=>0];
        $r=$conn->query("SELECT severity,COUNT(*) c FROM nm_health_predictions WHERE status='open' AND severity IN('warn','crit') GROUP BY severity");
        while($r&&$x=$r->fetch_assoc()) $out[$x['severity']]=(int)$x['c'];
        return $out;
    }
}
